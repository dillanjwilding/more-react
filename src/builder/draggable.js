import { Children, cloneElement } from 'react'

export const Draggable = ({ children }) => {
  const onDragStart = (e, component) => {
    e.dataTransfer.setData('text/plain', component)
  }
  return Children.map(children, (child, index) => {
    return cloneElement(child, {
      key: index, // not best practices
      draggable: true,
      onDragStart: e => onDragStart(e, child.props.name) // child.type.displayName
    })
  })
}
