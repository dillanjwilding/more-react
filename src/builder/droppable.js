// import 'bootstrap/dist/css/bootstrap.min.css'
import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

export const Droppable = ({ components }) => {
  const [values, setValues] = useState({ loading: true, items: [] }) // items = inputs
  useEffect(() => {
    setValues(state => ({ ...state, loading: false }))
  }, [])
  /* useEffect(() => {
    console.log(values.items)
  }, [values]) */
  const onDragOver = e => {
    e.preventDefault() // ?
  }
  // @todo: onDragOut or equivalent
  const onDrop = e => {
    const element = e.dataTransfer.getData('text')
    setValues(state => ({ ...state, items: [...values.items, element] }))
  }
  // @todo: This seems overly implemented for my forms; make more abstract
  return (
    <div onDragOver={onDragOver} onDrop={onDrop} className='col'>
      {values.items.map((name, i) => {
        const Component = components[name]
        // @todo: Figure out the UI/UX for getting the name of this component
        // The easiest would be to have a dialog before adding and rendering it
        if (Component.displayName === 'Submit') { // @todo: Fix this
          return <Component />
        }
        return <Component key={name + '-' + i}>Test</Component>
        // console.log(item)
        // return <div>{item}</div>
      })}
    </div>)
}
Droppable.propTypes = {
  components: PropTypes.object
}
