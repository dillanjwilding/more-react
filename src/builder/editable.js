import React, { Children, cloneElement, useState } from 'react'

export const Editable = ({ editing: edit = false, children, ...props }) => {
  const [editing, setEditing] = useState(edit)
  const [value, setValue] = useState('')
  return editing ? (
    <>
      <input onChange={e => setValue(e.target.value)} />
      <button onClick={() => setEditing(false)} />
    </>
  ) : (
    Children.map(children, (child, index) => {
      return cloneElement(child, {
        key: index, // not the best practice
        value
      })
    })
  )
}
