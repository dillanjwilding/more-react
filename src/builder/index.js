import 'core-js'
import 'regenerator-runtime/runtime'

import 'bootstrap/dist/css/bootstrap.min.css'

export * from './builder.js'
export * from './legend.js'
export * from './editable.js'
