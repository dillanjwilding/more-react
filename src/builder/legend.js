import React, { Children } from 'react'

import { Draggable } from './draggable.js'

// @todo: not best key practice; fix it
export const Legend = ({ children }) =>
  Children.map(children, (child, index) =>
    <Draggable key={index}>
      <div name={child.type.displayName}>
        {child.type.displayName}
      </div>
    </Draggable>
  )
