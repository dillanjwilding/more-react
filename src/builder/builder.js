import React, { useEffect } from 'react' // import { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import 'bootstrap/dist/css/bootstrap.min.css'

import * as fields from '../form/fields.js'
import { Droppable } from './droppable.js'
import { Legend } from './legend.js'

// @todo: Save to and load from JSON
// @todo: Add layout and formatting
/**
 * Take an object of displayed name => JSX and create a Legend containing display names
 * with the property/UX that can drag onto the Form but when dropped it adds the
 * corresponding JSX to the form.
 *
 * Or preferably/alternatively use the name or displayName of the component for the label.
 */
export const Builder = ({ children, render, autoSave = false }) => {
  // Does Builder need to know state so that it can save and load it?
  // const [values, setValues] = useState({ loading: true, items: [] }) // items = inputs
  const load = () => {}
  // const save = () => {}
  useEffect(() => {
    if (load) {
      load() // how to load into values?
    }
  }, [])
  /* useEffect(() => {
    if (autoSave) {

    }
  }, [values]) */
  const add = () => {

  }
  const remove = () => {

  }
  if (render) {
    return render({
      form: <Droppable components={fields} add={add} remove={remove} />,
      legend: <Legend>{children}</Legend>
    })
  } else {
    return (
      <div className='container-fluid'>
        <div className='row'>
          <Droppable components={fields} add={add} remove={remove} className='h-100' />
          <div className='col-2 h-100'>
            <Legend>{children}</Legend>
          </div>
        </div>
      </div>)
  }
}
Builder.propTypes = {
  render: PropTypes.func,
  autoSave: PropTypes.func
}
