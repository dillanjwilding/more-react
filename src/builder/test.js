/* global describe, test, expect */
import React from 'react'
import { shallow } from 'enzyme'

import { Builder } from './builder.js'

describe('Test builder', () => {
  test('renders correctly', () => {
    expect(shallow(<Builder />)).toMatchSnapshot()
  })
})

describe('Test [builder parts]', () => {

})
