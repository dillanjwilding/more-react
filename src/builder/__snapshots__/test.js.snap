// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test builder renders correctly 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <Droppable
      add={[Function]}
      className="h-100"
      components={
        Object {
          "Area": [Function],
          "Dropdown": [Function],
          "Email": [Function],
          "Field": [Function],
          "Option": [Function],
          "Password": [Function],
          "Submit": [Function],
        }
      }
      remove={[Function]}
    />
    <div
      className="col-2 h-100"
    >
      <Legend />
    </div>
  </div>
</div>
`;
